Mathlib/NumberTheory/NumberField/Completion.lean

```diff
@@ -0,0 +1,339 @@
+  when `v` is real.
+
+## Tags
+number field, embeddings, infinite places, completion
+-/
+noncomputable section
+
+/-- Type synonym equipping a semiring with an absolute value. -/
+@[nolint unusedArguments]
+def WithAbs {R S : Type*} [Semiring R] [OrderedSemiring S] :
+    AbsoluteValue R S → Type _ := fun _ => R
+
+namespace WithAbs
+
+variable {K : Type*} [Field K] (v : AbsoluteValue K ℝ)
+
+instance normedField : NormedField (WithAbs v) :=
+  v.normedField
+
+variable {L : Type*} [NormedField L] {f : WithAbs v →+* L} {v}
+
+/-- If the absolute value `v` factors through an embedding `f` into a normed field, then
+the distance associated to the absolute value also factors through `f`. -/
+theorem dist_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective)
+    (x y : WithAbs v) :
+    dist x y = dist (f x) (f y) := by
+  rw [(normedField v).dist_eq, (inferInstanceAs <| NormedField L).dist_eq,
+    ← f.map_sub, h]
+  rfl
+
+instance : Inhabited (WithAbs v) := ⟨0⟩
+
+/-- If the absolute value `v` factors through an embedding `f` into a normed field, then
+the pseudo metric space associated to the absolute value is the same as the pseudo metric space
+induced by `f`. -/
+theorem pseudoMetricSpace_induced_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) :
+    (normedField v).toPseudoMetricSpace = PseudoMetricSpace.induced f inferInstance := by
+  ext
+  exact dist_of_comp h _ _
+
+/-- If the absolute value `v` factors through an embedding `f` into a normed field, then
+the uniform structure associated to the absolute value is the same as the uniform structure
+induced by `f`. -/
+theorem uniformSpace_eq_comap_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) :
+    (normedField v).toUniformSpace = UniformSpace.comap f inferInstance := by
+  rw [pseudoMetricSpace_induced_of_comp h]
+  rfl
+
+/-- If the absolute value `v` factors through an embedding `f` into a normed field, then
+`f` is uniform inducing. -/
+theorem uniformInducing_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) :
+    UniformInducing f :=
+  uniformInducing_iff_uniformSpace.2 (Eq.symm (uniformSpace_eq_comap_of_comp h))
+
+/-- If the absolute value `v` factors through an embedding `f` into a normed field, then
+`f` is an isometry. -/
+theorem isometry_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) :
+    Isometry f :=
+  Isometry.of_dist_eq <| fun x y => by rw [pseudoMetricSpace_induced_of_comp h]; rfl
+
+end WithAbs
+
+namespace AbsoluteValue
+
+open WithAbs
+
+variable {K : Type*} [Field K] (v : AbsoluteValue K ℝ)
+
+/-- The completion of a field with respect to a real absolute value. -/
+abbrev completion := UniformSpace.Completion (WithAbs v)
+
+namespace Completion
+
+instance : NormedRing v.completion :=
```

Comment:
Why do you spell the instances below as such? They can automatically be inferred by type-class inference (if you replace the proof by `inferInstance`, lean is happy).

Reply:
Yes these can mostly be removed, except `Coe`.